### PR TITLE
Outlet/refactor

### PR DESCRIPTION
**Pull request by aleemont1**
I think I'll need some help w/ this.

### CODE REVIEW

File: src/main/java/homer/model/outlets/Outlet.java
Comment (michelececcacci) [inline src/main/java/homer/model/outlets/Outlet.java:53]: what is the point of casting here? `OutletState` should implement `DeviceState` and be its own class. Check out my commits to see how i did it

**Review by michelececcacci — CHANGES_REQUESTED** [review]
Check previous commits to see how TemperatureChangerState is implemented

**Comment by michelececcacci** [comment]
@smagnaniuni  can you take a look? i think it should be fine

**Comment by aleemont1** [comment]
For next times, please document better big changes like this one, I can't loose time go seeing what you've done in other classes, try to figure it out by myself by interpreting your code, and the apply changes to mine hoping not to break anything. 
Use more comments and write a proper guide to implementation if it should ever happen again to make a big refactor like this one. 
Lmk if I missed something in the last commits.

**Comment by michelececcacci** [comment]
>For next times, please document better big changes like this one

Ok, noted. I thought my commits were pretty self-explanatory

**Comment by aleemont1** [comment]
Also, CI is not passing anymore. I can't even test my classes due to errors introduced with the latest changes, when are these going to be fixed? 
I don't think they should have been merged in dev before fixing the errors. Let's not mess up all the work we've done for doing things too fast, please.

**Comment by michelececcacci** [comment]
>Also, CI is not passing anymore. I can't even test my classes due to errors introduced with the latest changes, when are these going to be fixed?

As soon as we get everything merged.

**Comment by aleemont1** [comment]
> > For next times, please document better big changes like this one
> 
> Ok, noted. I thought my commits were pretty self-explanatory

I wouldn't look at commits as an implementation guide. Also it shouldn't be my work to cherry pick all the changes made for implementing something you introduced, just a few comments in the code or in the PR, that's it. 
LMK if I missed anything in this PR and then i'll merge it.

File: src/main/java/homer/controller/impl/electricalmeter/ElectricalMeterImpl.java
Comment (aleemont1) [inline src/main/java/homer/controller/impl/electricalmeter/ElectricalMeterImpl.java:72]: How can I implement this using Streams?

File: src/main/java/homer/model/outlets/Outlet.java
Comment (aleemont1) [inline src/main/java/homer/model/outlets/Outlet.java:73]: Is this right?

File: src/main/java/homer/model/outlets/Outlet.java
Comment (aleemont1) [inline src/main/java/homer/model/outlets/Outlet.java:90]: Also this, or should I
`OutletState energy = (OutletState) this.getState()` ?

File: src/main/java/homer/controller/impl/electricalmeter/ElectricalMeterImpl.java
Comment (smagnaniuni) [reply src/main/java/homer/controller/impl/electricalmeter/ElectricalMeterImpl.java:72]: Probly something like
this.getOutlets().stream().map(e -> e.getState()).map(e -> e.getPower())

then use filter and map again to get the values, then turn it into a DoubleStream and you can use .sum()

File: src/main/java/homer/model/outlets/Outlet.java
Comment (smagnaniuni) [reply src/main/java/homer/model/outlets/Outlet.java]: Depends what is this supposed to do?

Comment (smagnaniuni) [reply src/main/java/homer/model/outlets/Outlet.java]: Also you can save yourself the headache and keep the internal implementation untouched, technically there should only be changes in getState() and setState(), just creating and retrieving the appropriate DeviceState implementation for your needs.

Comment (michelececcacci) [reply src/main/java/homer/model/outlets/Outlet.java]: > Also you can save yourself the headache and keep the internal implementation untouched, technically there should only be changes in getState() and setState(), just creating and retrieving the appropriate DeviceState implementation for your needs.

100%, but let's not get too hung up on some private method's implementation. If observable behaviour is fine, and we don't have performance issues, it's good enough.

File: src/main/java/homer/model/outlets/OutletFactory.java
Comment (michelececcacci) [inline src/main/java/homer/model/outlets/OutletFactory.java:26]: minor, but you can just chain method calls here

**Review by michelececcacci — CHANGES_REQUESTED** [review]

File: src/main/java/homer/model/outlets/Outlet.java
Comment (aleemont1) [reply src/main/java/homer/model/outlets/Outlet.java:73]: Yep ur right

**Comment by michelececcacci** [comment]
> Also, CI is not passing anymore. I can't even test my classes due to errors introduced with the latest changes, when are these going to be fixed?

actually, the changes you are making right now should fix all build errors, and allow us to test the classes again

**Comment by aleemont1** [comment]
> > Also, CI is not passing anymore. I can't even test my classes due to errors introduced with the latest changes, when are these going to be fixed?
> 
> actually, the changes you are making right now should fix all build errors, and allow us to test the classes again. You can always run unit tests with `./gradlew test`, even when checkstyle fails

Not if building is failing.
Btw, can I merge this?

**Comment by michelececcacci** [comment]
>Not if building is failing.
Btw, can I merge this?

are you synced with upstream/refactor? in that case, it should work. I fixed all of my classes, and i think @smagnaniuni  did too. So if you also fixed them, it should build

**Comment by michelececcacci** [comment]
@aleemont1 after looking at the latest CI log on this PR, it fails because of an error in your class. Please fix it.

**Comment by aleemont1** [comment]
> @aleemont1 after looking at the latest CI log on this PR, the build fails because of an error in your class. Please fix it. https://github.com/progetto-oop-22-23/OOP22-HOMER/actions/runs/4525293064/jobs/7969686923?pr=141

I get an error on the getState return type, so the problem is with the Device interface, which exptects a generic as return type, since the generic type in Outlet is Double, it's expecting a Double as return type, pretty sure not my fault.

**Comment by smagnaniuni** [comment]
The problem is you're using AdjustableDevice<Double>, where instead of Double it should take your DeviceState implementation.

@michelececcacci 
I think we may need to put <S extends DeviceState> in the generic of the device interfaces, so that it's not possible to mistakenly use "primitive" types?

**Comment by michelececcacci** [comment]
>The problem is you're using AdjustableDevice<Double>, where instead of Double it should take your DeviceState implementation.

My temperaturechanger also uses AdjustableDevice<Double>, but it works. Definitely an implementation problem on his side.

**Comment by smagnaniuni** [comment]
> > The problem is you're using AdjustableDevice, where instead of Double it should take your DeviceState implementation.
> 
> My temperaturechanger also uses AdjustableDevice, but it works. Definitely an implementation problem on his side.

Yeah I mean he used Double instead of something implementing DeviceState, github markdown cut out the other part.

**Comment by michelececcacci** [comment]
now we have checkstyle fails, so we can merge, as far as i'm concerned.

**Review by michelececcacci — APPROVED** [review]

**Comment by aleemont1** [comment]
> > > The problem is you're using AdjustableDevice, where instead of Double it should take your DeviceState implementation.
> > 
> > 
> > My temperaturechanger also uses AdjustableDevice, but it works. Definitely an implementation problem on his side.
> 
> Yeah I mean he used Double instead of something implementing DeviceState, github markdown cut out the other part.

Is it ok now?

**Comment by michelececcacci** [comment]
yeah, i just asked for simone's opinion. but i feel like the sooner we merge, the better.

**Comment by aleemont1** [comment]
> yeah, i just asked for simone's opinion. but i feel like the sooner we merge, the better.

Okie, will wait for his response then.

File: src/main/java/homer/model/outlets/OutletState.java
Comment (smagnaniuni) [inline src/main/java/homer/model/outlets/OutletState.java:6]: probably missing javadoc

**Review by smagnaniuni — APPROVED** [review]
lgtm

**Pull request merged**
